src/Language/PureScript/CoreFn/CSE.hs

```diff
@@ -0,0 +1,425 @@
+import Language.PureScript.Names
+import Language.PureScript.PSString (decodeString)
+
+-- |
+-- `discuss f m` is an action that listens to the output of `m`, passes that
+-- and its value through `f`, and uses (only) the value of the result to set
+-- the new value and output. (Any output produced via the monad in `f` is
+-- ignored, though other monadic effects will hold.)
+--
+discuss :: MonadWriter w m => ((a, w) -> m (b, w)) -> m a -> m b
+discuss f = pass . fmap (second const) . (>>= f) . listen
+
+-- |
+-- Modify the target of an optic in the state with a monadic computation that
+-- returns some extra information of type `r` in a tuple.
+--
+-- I would prefer that this be a named function, but I don't know what to name
+-- it. I went with symbols instead because the function that this operator most
+-- resembles is `(%%=)`, which doesn't have a textual name as far as I know.
+-- Compare the following (approximate) types:
+--
+-- @
+-- (%%=)  :: MonadState s m => Lens s s a b -> (a ->   (r, b)) -> m r
+-- (%%<~) :: MonadState s m => Lens s s a b -> (a -> m (r, b)) -> m r
+-- @
+--
+-- Replacing the `=` with `<~` was inspired by analogy with the following pair:
+--
+-- @
+-- (.=) :: MonadState s m => Lens s s a b ->   b -> m ()
+-- (<~) :: MonadState s m => Lens s s a b -> m b -> m ()
+-- @
+--
+-- I regret any confusion that ensues.
+--
+-- Note that there are two interpretations that could reasonably be expected
+-- for this type.
+--
+-- @
+-- (%%<~) :: MonadState s m => Lens s s a b -> (a -> m (r, b)) -> m r
+-- @
+--
+-- One is:
+-- * Get the focused `a` value from the monad
+-- * Run the computation
+-- * Get the new state from the returned monad
+-- * Take the returned `b` value and set it in the new state
+--
+-- The other is:
+-- * Get the focused `a` value from the monad
+-- * Run the computation
+-- * Take the returned `b` value and set it in the *original* state
+-- * Put the result into the returned monad
+--
+-- This operator corresponds to the second interpretation. The purpose of this,
+-- and part of the purpose of having this operator at all instead of composing
+-- simpler operators, is to enable using the lens only once (on the original
+-- state) instead of twice (for a get and a set on different states).
+--
+(%%<~)
+  :: MonadState s m
+  => ((a -> Compose m ((,) r) b) -> s -> Compose m ((,) r) s)
+     -- ^ please read as Lens s s a b
+  -> (a -> m (r, b))
+  -> m r
+l %%<~ f = get >>= getCompose . l (Compose . f) >>= state . const
+infix 4 %%<~
+
+-- |
+-- A PluralityMap is like a weaker multiset: like a multiset, it holds a
+-- number of values, but instead of keeping track of their exact counts, it
+-- only records whether there is one (False) or more than one (True).
+--
+newtype PluralityMap k = PluralityMap { getPluralityMap :: M.Map k Bool }
+
+instance Ord k => Semigroup (PluralityMap k) where
+  PluralityMap l <> PluralityMap r =
+    PluralityMap $ M.mapWithKey (\k -> (|| k `M.member` r)) l `M.union` r
```

Comment:
Just to clarify, is this
```hs
PluralityMap $ (M.mapWithKey (\k -> (|| k `M.member` r)) l) `M.union` r
```
or this?
```
PluralityMap $ M.mapWithKey (\k -> (|| k `M.member` r)) (l `M.union` r)
```
I believe it's the second one.

Reply:
I think it's the first, actually. Prefix function application has higher precedence than all operators, including backticked identifiers.

Reply:
:smile: Maybe we should add parenthesis here just to ensure clarity? Though I don't think there's any difference between the two.

Reply:
Yeah, the only difference would be performance, if any. I think I'd have to fight with hlint about those parentheses but I'll try and see.

Reply:
(Oh, I shouldn't have tried to comment on this while going to bed. There is an important difference between the two: if `l = PluralityMap M.empty` and `r = PluralityMap $ M.singleton 0 False`, then in the first, `l <> r == PluralityMap $ M.singleton 0 False`, but in the second, `l <> r == PluralityMap $ M.singleton 0 True`.)